File: GeometricObjects/Export.py
from collections import defaultdict
import os
from PIL import Image
import numpy as np
from Triangle import *




def stripMaterialInformation(lines):  # strip any preexisting material information
    result = []
    for line in lines:
        if len(line) >= 6 and line[0:6] in ["mtllib", "usemtl"]:
            pass
        else:
            result.append(line)
    return result


def getMatLines(mtl_name, location):
    return [
        f"\nnewmtl mat_{mtl_name}",
        "Ns 250.000000",
        "Ka 1.000000 1.000000 1.000000",
        "Kd 0.097925 0.065680 0.800000",
        "Ks 0.500000 0.500000 0.500000",
        "Ke 0.000000 0.000000 0.000000",
        "Ni 1.450000",
        "d 1.000000",
        "illum 2",
        f"map_Kd ./images/{location}.png",
    ]


def writeNewMtl(OUT_LOCATION, MTL_FILE_NAME, objects):
    with open(OUT_LOCATION + MTL_FILE_NAME + ".mtl", "w+") as f:
        for object in objects:
            f.writelines(line + "\n" for line in getMatLines(object, f"mat_{object}"))


def writeNewObj(OUT_LOCATION, FILENAME, newFile):
    with open(OUT_LOCATION + FILENAME + "_textured.obj", "w+") as f:
        f.write(newFile)


class hit:
    def __init__(self, name, hit_num, coords) -> None:
        self.name = name
        self.hit_num = hit_num
        self.x, self.y = coords

    def __repr__(self) -> str:
        return f"{self.name} - {self.hit_num}: <{self.x}, {self.y}>"





def writeImages(objects: list[TriObject]):
    images = {}
    print("Writing images")
    for i, obj in enumerate(objects):
        print(f"Writing image for {obj.name} ({i+1}/{len(objects)})")
        
        img = np.zeros((obj.resolution, obj.resolution, 3), dtype=np.uint8)
        for i in range(obj.resolution):
            for j in range(obj.resolution):
                img[i][j] = obj.texture[i][j].getColor()
                #print(img[i][j])
        #print(img)
        image = Image.fromarray(img)  
        image.save(f"./Textured/images/mat_{obj.name}.png")

def main(FILENAME, objs):
    LOCATION = "./"
    OUT_LOCATION = "./Textured/"
    # FILENAME = "ReflectionTestWIthCube"
    MTL_FILE_NAME = None or FILENAME + "_MTL"
    newFile = ""  # string with content of modified file
    lines = None
    with open(LOCATION + FILENAME + ".obj", "r") as f:
        lines = f.readlines()

    lines = stripMaterialInformation(lines)

    lines.insert(0, f"mtllib {MTL_FILE_NAME}.mtl\n")

    objects = []

    curObject = None
    doJoin = True
    for line in lines:
        if line[0] == "o":
            print("object")
            curObject = line.split(" ")[1]
            objects.append(curObject.strip())
            doJoin = True
        elif line[0] == "s":
            newFile += line
            newFile += f"usemtl mat_{curObject}\n"
            doJoin = False
        elif len(line) >= 6 and line[0:6] == "usemtl":
            doJoin = False
        if doJoin:
            newFile += line
        else:
            doJoin = True
    print("Objs:", objects)
    writeNewMtl(OUT_LOCATION, MTL_FILE_NAME, objects)
    writeNewObj(OUT_LOCATION, FILENAME, newFile)
    writeImages(objs)


import sys

File: rust/src/export.rs
use crate::triangle::*;
use crate::vector::*;
use image::{DynamicImage, ImageBuffer, Rgb};
use std::collections::HashMap;
use std::fs;
use std::fs::File;
use std::io::{self, BufRead, Write};
use std::path::Path;
use std::rc::Rc;
struct Color {
    r: u8,
    g: u8,
    b: u8,
}

impl Color {
    pub fn get_color(&self) -> [u8; 3] {
        [self.r, self.g, self.b]
    }
}

fn strip_material_information(lines: Vec<String>) -> Vec<String> {
    lines
        .into_iter()
        .filter(|line| line.len() >= 6 && &line[0..6] != "mtllib" && &line[0..6] != "usemtl")
        .collect()
}

fn get_mat_lines(mtl_name: &str, location: &str) -> Vec<String> {
    vec![
        format!("\nnewmtl mat_{}", mtl_name),
        "Ns 250.000000".to_string(),
        "Ka 1.000000 1.000000 1.000000".to_string(),
        "Kd 0.097925 0.065680 0.800000".to_string(),
        "Ks 0.500000 0.500000 0.500000".to_string(),
        "Ke 0.000000 0.000000 0.000000".to_string(),
        "Ni 1.450000".to_string(),
        "d 1.000000".to_string(),
        "illum 2".to_string(),
        format!("map_Kd ./images/{}.png", location),
    ]
}

fn write_new_mtl(out_location: &str, mtl_file_name: &str, objects: &Vec<String>) -> io::Result<()> {
    let path = format!("{}{}.mtl", out_location, mtl_file_name);
    let mut file = File::create(path)?;
    for object in objects {
        for line in get_mat_lines(object, &format!("mat_{}", object)) {
            writeln!(file, "{}", line)?;
        }
    }
    Ok(())
}

fn write_new_obj(out_location: &str, file_name: &str, new_file: String) -> io::Result<()> {
    let path = format!("{}{}_textured.obj", out_location, file_name);
    fs::write(path, new_file)?;
    Ok(())
}

fn get_color(status: u8) -> (u8, u8, u8) {
    return match status {
        0 => (200, 200, 200),
        1 => (255, 0, 0),
        2 => (0, 255, 0),
        _ => (0, 0, 0),
    };
}

fn write_images(objects: Vec<Rc<TriObject>>) {
    println!("Writing images");
    for (i, obj) in objects.iter().enumerate() {
        println!(
            "Writing image for {} ({}/{})",
            obj.name,
            i + 1,
            objects.len()
        );

        let mut img = ImageBuffer::new(obj.resolution as u32, obj.resolution as u32);
        for x in 0..obj.resolution {
            for y in 0..obj.resolution {
                let st = obj.texture[x as usize][y as usize];
                let color = get_color(st);
                img.put_pixel(x as u32, y as u32, Rgb(color.into()));
            }
        }

        let dyn_image = DynamicImage::ImageRgb8(img);
        dyn_image
            .save(format!("./Textured/images/mat_{}.png", obj.name))
            .unwrap();
    }
}

pub fn export(filename: &str, objs: Vec<Rc<TriObject>>) {
    let out_location = "./Textured/";
    //let filename = "ReflectionTestWithCube"; // Adjust filename as necessary
    let mtl_file_name = format!("{}_MTL", filename);

    let path = format!("{}.obj", filename);
    println!("Opening {}", path);
    let input = File::open(&path).expect("Unable to open file");
    let buffered = io::BufReader::new(input);

    let mut lines: Vec<String> = buffered
        .lines()
        .map(|line| line.expect("Could not parse line"))
        .collect();

    lines = strip_material_information(lines);

    lines.insert(0, format!("mtllib {}.mtl\n", mtl_file_name));

    let mut objects = vec![];
    let mut cur_object = String::new();
    let mut new_file = String::new();
    let mut do_join = true;

    for line in lines.iter() {
        if line.chars().next().unwrap() == 'o' {
            cur_object = line.split_whitespace().nth(1).unwrap().to_string();
            objects.push(cur_object.clone().trim().to_string());
            do_join = true;
        } else if line.chars().next().unwrap() == 's' {
            new_file.push_str(line);
            new_file.push_str(&format!("usemtl mat_{}\n", cur_object));
            do_join = false;
        } else if line.len() >= 6 && &line[0..6] == "usemtl" {
            do_join = false;
        }
        if do_join {
            new_file.push_str(line);
        } else {
            do_join = true;
        }
    }

    println!("Objs: {:?}", objects);
    println!("{}", out_location);
    write_new_mtl(out_location, &mtl_file_name, &objects).expect("Failed to write new MTL file");

    write_new_obj(out_location, filename, new_file).expect("Failed to write new OBJ file");

    // Here, replace `Vec::new()` with your vector of TriObject instances.
    // Ensure to define and populate your TriObject instances with the necessary data before this point.
    write_images(objs);
}
